### ISSUE TITLE
[CURATOR-275] Allow service instances to be disabled temporarily

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#795]
<p>Allow service instances to be temporarily removed from discovery (when using the Service Discovery recipe). </p>

<p>This can be achieved with a new 'enabled' flag that is included as part of the service instance node data in Zookeeper. Updating this flag via e.g. Exhibitor will allow an instance to be disabled. When the flag is omitted, the service is assumed to be enabled, hence this change is backward compatible.</p>

---
<details><summary><i>Originally reported by <a title="Joe Littlejohn" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=japher">japher</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-275" target="_blank">Allow service instances to be disabled temporarily</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jordan Zimmerman" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-01-24T02:08:30+00:00
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user serranom commented on the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/116#issuecomment-153188362" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116#issuecomment-153188362</a></p>

<p>    why not just unregister and then re-register the service?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user serranom commented on the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/116#issuecomment-153753850" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116#issuecomment-153753850</a></p>

<p>    Makes sense now.  Thanks.  I do wonder how you would envision this being<br/>
    used at the API level.  Changing data just at the ZK level implies internal<br/>
    knowledge of the znode layout which doesn't seem to be intended to be<br/>
    exposed by the service-discovery framework.</p>


<p>    On Mon, Nov 2, 2015 at 7:20 PM Joe Littlejohn <notifications@github.com><br/>
    wrote:</p>

<p>    > There's some more discussion about this on the mailing list<br/>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user joelittlejohn commented on the pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/116#issuecomment-154554250" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116#issuecomment-154554250</a></p>

<p>    It's a very good point @serranom. </p>

<p>    There's nothing stopping someone from doing this via the API (I don't think) since you can simply create a new ServiceInstance (based on the field values of the instance you want to disable) but pass `false` as the enabled value, then use <a href="#updateService(ServiceInstance)" target="_blank" rel="noopener">ServiceDiscovery#updateService(ServiceInstance)</a>(<a href="https://github.com/joelittlejohn/curator/blob/4776a2daacb3a776e9dd0bdc214ed4987fef96c9/curator-x-discovery/src/main/java/org/apache/curator/x/discovery/ServiceDiscovery.java#L48" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/joelittlejohn/curator/blob/4776a2daacb3a776e9dd0bdc214ed4987fef96c9/curator-x-discovery/src/main/java/org/apache/curator/x/discovery/ServiceDiscovery.java#L48</a>) to apply the change. This is the kind of thing a dashboard/tool (that has no knowledge of the znode layout) could do. So in this way, all ServiceInstance fields are always updatable via the API.</p>

<p>    I saw this as an administrative task that (as the least worst option) allows an update to the znode to disable an instance. I thought of this as a useful administrative hook but not necessarily an operation that needs to part of the core API. If we did want <em>improved</em> support for this in the API it would be easy to add this e.g. </p>

<p>    `ServiceDiscovery#enable(ServiceInstance)` and `ServiceDiscovery#disable(ServiceInstance)`</p>

<p>    or</p>

<p>    `ServiceInstance#enable()` and `ServiceInstance#disable()`<br/>
    (followed by `ServiceDiscovery#updateService(ServiceInstance)`)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/curator/pull/116" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>My apologies that this took so long to get reviewed. We really need more committers on the project.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user joelittlejohn commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/116" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116</a></p>

<p>    Thanks to whoever merged this! :+1:</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Randgalt commented on the issue:</p>

<p>    <a href="https://github.com/apache/curator/pull/116" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/116</a></p>

<p>    You're welcome. I'm sorry it took so long. We need more Curator committers.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user Randgalt opened a pull request:</p>

<p>    <a href="https://github.com/apache/curator/pull/208" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/208</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-275" title="Allow service instances to be disabled temporarily" class="issue-link" data-issue-key="CURATOR-275"><del>CURATOR-275</del></a> UnrecognizedPropertyException: "enabled" incompatibility</p>

<p>    Here is a potential fix for the issue folks. Let me know what you think...</p>

<p>    For users who don't want the new "enabled" field serialized, you can create a JsonInstanceSerializer in compatibility mode:</p>

<p>    ```java<br/>
    new JsonInstanceSerializer(payloadClass, true)<br/>
    ```</p>

<p>    Internally, the JsonInstanceSerializer will not serialize the enabled field. There are tests to ensure that this will deserialize in old clients.</p>

<p>    Going forward, JsonInstanceSerializer now sets ignore fields in the ObjectMapper.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/apache/curator" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator</a> <a href="https://issues.apache.org/jira/browse/CURATOR-394" title="UnrecognizedPropertyException: "enabled" incompatibility" class="issue-link" data-issue-key="CURATOR-394"><del>CURATOR-394</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/curator/pull/208.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/curator/pull/208.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #208</p>

<hr />
<p>commit 070f1c9e4dae947ba21b015ea5027c3d20d170bd<br/>
Author: randgalt <randgalt@apache.org><br/>
Date:   2017-03-25T00:22:48Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/CURATOR-275" title="Allow service instances to be disabled temporarily" class="issue-link" data-issue-key="CURATOR-275"><del>CURATOR-275</del></a> introduced a new field into ServiceInstance. This caused a potential UnrecognizedPropertyException in older clients that read newly serialized ServiceInstances. Added an alternate ctor to JsonInstanceSerializer with a compatibleSerializationMode option. when set to true, the new enabled field of ServiceInstance is not serialized.</p>

<hr />
